File: honjo-masamune/honjo/src/lexer.ts
// Honjo Masamune — Lexer
// Implements the lexical structure of the language spec (§3 "Lexical structure").
// UTF-8 source, `.hj`. `--` line comments; whitespace is insignificant.

export type TokKind =
  | "kw"        // keyword
  | "ident"     // identifier
  | "num"       // numeric literal
  | "str"       // string literal
  | "op"        // operator / punctuation
  | "eof";

export interface Token {
  kind: TokKind;
  value: string;
  line: number;
  col: number;
}

// Keyword set from spec §3 (Definition: Token classes).
export const KEYWORDS = new Set<string>([
  "floor", "cut", "close", "track", "until", "yield", "when", "do",
  "emit", "observe", "in", "as", "let", "medium", "converge", "diverge",
  "with", "by", "import", "module", "export", "assert", "reps",
]);

// Multi-char operators, longest first so the scanner is greedy.
const MULTI_OPS = [":=", ">=", "<=", "==", "->"];
const SINGLE_OPS = new Set<string>([
  "~", "(", ")", "[", "]", "{", "}", ",", ":", ".", ">", "<", "#",
]);

export class LexError extends Error {
  constructor(msg: string, public line: number, public col: number) {
    super(`lex error at ${line}:${col}: ${msg}`);
    this.name = "LexError";
  }
}

const isIdentStart = (c: string) => /[A-Za-z_]/.test(c);
const isIdentPart = (c: string) => /[A-Za-z0-9_]/.test(c);
const isDigit = (c: string) => c >= "0" && c <= "9";

export function lex(src: string): Token[] {
  const toks: Token[] = [];
  let i = 0;
  let line = 1;
  let col = 1;
  const n = src.length;

  const peek = (k = 0) => src[i + k] ?? "";
  const adv = () => {
    const c = src[i++];
    if (c === "\n") { line++; col = 1; } else { col++; }
    return c;
  };

  while (i < n) {
    const c = peek();

    // whitespace
    if (c === " " || c === "\t" || c === "\r" || c === "\n") { adv(); continue; }

    // line comment  --  ...  to end of line
    if (c === "-" && peek(1) === "-") {
      while (i < n && peek() !== "\n") adv();
      continue;
    }

    const startLine = line, startCol = col;

    // string literal
    if (c === '"') {
      adv();
      let s = "";
      while (i < n && peek() !== '"') {
        const ch = adv();
        if (ch === "\\") {
          const e = adv();
          s += e === "n" ? "\n" : e === "t" ? "\t" : e;
        } else {
          s += ch;
        }
      }
      if (i >= n) throw new LexError("unterminated string", startLine, startCol);
      adv(); // closing quote
      toks.push({ kind: "str", value: s, line: startLine, col: startCol });
      continue;
    }

    // number: [0-9]+(.[0-9]+)?([eE][+-]?[0-9]+)?
    if (isDigit(c) || (c === "." && isDigit(peek(1)))) {
      let s = "";
      while (isDigit(peek())) s += adv();
      if (peek() === ".") { s += adv(); while (isDigit(peek())) s += adv(); }
      if (peek() === "e" || peek() === "E") {
        s += adv();
        if (peek() === "+" || peek() === "-") s += adv();
        if (!isDigit(peek())) throw new LexError("malformed exponent", startLine, startCol);
        while (isDigit(peek())) s += adv();
      }
      toks.push({ kind: "num", value: s, line: startLine, col: startCol });
      continue;
    }

    // identifier / keyword
    if (isIdentStart(c)) {
      let s = "";
      while (isIdentPart(peek())) s += adv();
      toks.push({ kind: KEYWORDS.has(s) ? "kw" : "ident", value: s, line: startLine, col: startCol });
      continue;
    }

    // multi-char operator
    const two = c + peek(1);
    if (MULTI_OPS.includes(two)) {
      adv(); adv();
      toks.push({ kind: "op", value: two, line: startLine, col: startCol });
      continue;
    }

    // single-char operator
    if (SINGLE_OPS.has(c)) {
      adv();
      toks.push({ kind: "op", value: c, line: startLine, col: startCol });
      continue;
    }

    throw new LexError(`unexpected character '${c}'`, startLine, startCol);
  }

  toks.push({ kind: "eof", value: "", line, col });
  return toks;
}

File: honjo-masamune/honjo-rs/src/lexer.rs
//! Honjo Masamune — Lexer (spec §3). UTF-8 `.hj`, `--` line comments.

#[derive(Debug, Clone, PartialEq)]
pub enum TokKind {
    Kw,
    Ident,
    Num,
    Str,
    Op,
    Eof,
}

#[derive(Debug, Clone)]
pub struct Token {
    pub kind: TokKind,
    pub value: String,
    pub line: usize,
    pub col: usize,
}

pub const KEYWORDS: &[&str] = &[
    "floor", "cut", "close", "track", "until", "yield", "when", "do", "emit",
    "observe", "in", "as", "let", "medium", "converge", "diverge", "with",
    "by", "import", "module", "export", "assert", "reps",
];

const MULTI_OPS: &[&str] = &[":=", ">=", "<=", "==", "->"];
const SINGLE_OPS: &[char] =
    &['~', '(', ')', '[', ']', '{', '}', ',', ':', '.', '>', '<', '#'];

#[derive(Debug)]
pub struct LexError {
    pub msg: String,
    pub line: usize,
    pub col: usize,
}

impl std::fmt::Display for LexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "lex error at {}:{}: {}", self.line, self.col, self.msg)
    }
}

fn is_ident_start(c: char) -> bool {
    c.is_ascii_alphabetic() || c == '_'
}
fn is_ident_part(c: char) -> bool {
    c.is_ascii_alphanumeric() || c == '_'
}

pub fn lex(src: &str) -> Result<Vec<Token>, LexError> {
    let chars: Vec<char> = src.chars().collect();
    let n = chars.len();
    let mut i = 0usize;
    let mut line = 1usize;
    let mut col = 1usize;
    let mut toks = Vec::new();

    macro_rules! adv {
        () => {{
            let c = chars[i];
            i += 1;
            if c == '\n' {
                line += 1;
                col = 1;
            } else {
                col += 1;
            }
            c
        }};
    }

    while i < n {
        let c = chars[i];

        // whitespace
        if c == ' ' || c == '\t' || c == '\r' || c == '\n' {
            adv!();
            continue;
        }
        // line comment
        if c == '-' && i + 1 < n && chars[i + 1] == '-' {
            while i < n && chars[i] != '\n' {
                adv!();
            }
            continue;
        }

        let (sl, sc) = (line, col);

        // string
        if c == '"' {
            adv!();
            let mut s = String::new();
            while i < n && chars[i] != '"' {
                let ch = adv!();
                if ch == '\\' && i < n {
                    let e = adv!();
                    s.push(match e {
                        'n' => '\n',
                        't' => '\t',
                        other => other,
                    });
                } else {
                    s.push(ch);
                }
            }
            if i >= n {
                return Err(LexError { msg: "unterminated string".into(), line: sl, col: sc });
            }
            adv!(); // closing quote
            toks.push(Token { kind: TokKind::Str, value: s, line: sl, col: sc });
            continue;
        }

        // number
        if c.is_ascii_digit() || (c == '.' && i + 1 < n && chars[i + 1].is_ascii_digit()) {
            let mut s = String::new();
            while i < n && chars[i].is_ascii_digit() {
                s.push(adv!());
            }
            if i < n && chars[i] == '.' {
                s.push(adv!());
                while i < n && chars[i].is_ascii_digit() {
                    s.push(adv!());
                }
            }
            if i < n && (chars[i] == 'e' || chars[i] == 'E') {
                s.push(adv!());
                if i < n && (chars[i] == '+' || chars[i] == '-') {
                    s.push(adv!());
                }
                if i >= n || !chars[i].is_ascii_digit() {
                    return Err(LexError { msg: "malformed exponent".into(), line: sl, col: sc });
                }
                while i < n && chars[i].is_ascii_digit() {
                    s.push(adv!());
                }
            }
            toks.push(Token { kind: TokKind::Num, value: s, line: sl, col: sc });
            continue;
        }

        // identifier / keyword
        if is_ident_start(c) {
            let mut s = String::new();
            while i < n && is_ident_part(chars[i]) {
                s.push(adv!());
            }
            let kind = if KEYWORDS.contains(&s.as_str()) { TokKind::Kw } else { TokKind::Ident };
            toks.push(Token { kind, value: s, line: sl, col: sc });
            continue;
        }

        // multi-char operator
        if i + 1 < n {
            let two: String = [chars[i], chars[i + 1]].iter().collect();
            if MULTI_OPS.contains(&two.as_str()) {
                adv!();
                adv!();
                toks.push(Token { kind: TokKind::Op, value: two, line: sl, col: sc });
                continue;
            }
        }

        // single-char operator
        if SINGLE_OPS.contains(&c) {
            adv!();
            toks.push(Token { kind: TokKind::Op, value: c.to_string(), line: sl, col: sc });
            continue;
        }

        return Err(LexError { msg: format!("unexpected character '{}'", c), line: sl, col: sc });
    }

    toks.push(Token { kind: TokKind::Eof, value: String::new(), line, col });
    Ok(toks)
}
